darwin/dataset/upload_manager.py

```diff
@@ -1,528 +1,610 @@
+        self.data = kwargs
+        self._type_check(kwargs)
+
+    def _type_check(self, args):
+        self.data["filename"] = args.get("filename") or self.local_path.name
+        self.data["type"] =  args.get("type") or self._infer_type()  
+
+    def _infer_type(self):
+        extension = Path(self.data["filename"]).suffix
+        if is_image_extension_allowed(extension):
+            return "image"
+        if is_video_extension_allowed(extension):
+            return "video"
+        raise UnsupportedFileType(self.data["filename"])
+
+class UploadHandler:
+    def __init__(self, client, local_files, dataset_identifier):
+        self.client = client 
+        self.local_files = local_files
+        self.dataset_identifier = dataset_identifier
+
+        self.blocked_items, self.pending_items = request_upload(client, local_files, dataset_identifier)
+        print(self.blocked_items, self.pending_items)
+
+    @property
+    def pending_count(self):
+        return len(self.pending_items)
+
+    @property
+    def blocked_count(self):
+        return len(self.blocked_items)
+
+    def upload(self):
+        return _upload_files(self.client, self.local_files, self.pending_items, self.dataset_identifier.team_slug)
+    
+
+def request_upload(client: "Client", files: List[LocalFile], dataset_identifier):
+    upload_payload = { "items": [file.data for file in files] }
+    data = client.put(endpoint=f"/teams/{dataset_identifier.team_slug}/datasets/{dataset_identifier.dataset_slug}/data", payload=upload_payload, team=dataset_identifier.team_slug)
+    return data["blocked_items"], data["items"]
+
+# def add_files_to_dataset(client: "Client", files: List[LocalFile], dataset_slug: str, team_slug: str):
+#     upload_payload = { "items": [file.data for file in files] }
+#     data = client.put(endpoint=f"/teams/{team_slug}/datasets/{dataset_slug}/data", payload=upload_payload, team=team_slug)
+    
+#     print(data)
+
+#     if 'items' in data and data['items']:
+#         _upload_files(client, files, data['items'], team_slug)
+
+def _upload_files(client: "Client", files: List[LocalFile], items_pending_upload, team_slug: str):
+    file_lookup = { file.data["filename"]: file for file in files }
+    for item in items_pending_upload:
+        file = file_lookup.get(item['filename'])
+        if not file:
+            raise ValueError(f"Can not match {item['filename']} from payload with files to upload")
+        yield lambda: _upload_file(client, item["dataset_item_id"], file.local_path, team_slug=team_slug)
+
+def _upload_file(client: "Client", dataset_item_id: int, file_path: Path, team_slug: str):
+    sign_response = client.get(f"/dataset_items/{dataset_item_id}/sign_upload", team=team_slug)
+    signature = sign_response["signature"]
+    end_point = sign_response["postEndpoint"]
+
+    upload_response = requests.post("http:" + end_point, data=signature, files={"file": file_path.open("rb")})
+    upload_response.raise_for_status()
```

Comment:
If we raise here, does that get caught for us to print a more descriptive error message?